File: Extensions/Tessa.Extensions.Client/UI/PnrUKOrderUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrUKOrderUIExtension : CardUIExtension
    {
        private readonly ICardRepository cardRepository;
        public PnrUKOrderUIExtension(ICardRepository cardRepository)
        {
            this.cardRepository = cardRepository;
        }

        private async void SetUserDepartmentInfo(Guid? authorID, ICardModel cardModel)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
                Info =
                {
                    { "authorID", authorID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful)
            {
                cardModel.Card.Sections["PnrOrderUK"].Fields["DepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                cardModel.Card.Sections["PnrOrderUK"].Fields["DepartmentName"] = response.Info.Get<string>("Name");
                cardModel.Card.Sections["PnrOrderUK"].Fields["DepartmentIdx"] = response.Info.Get<string>("Index");
            }
        }

        private async void SetLegalEntityIndex(Guid organizationID, ICardModel cardModel)
        {
            // request к организации получить индекс ЮЛ
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetIndexLegalEntityRequestTypeID,
                Info =
                {
                    { "organizationID", organizationID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful)
            {
                //cardModel.Card.Sections["PnrOrderUK"].Fields["LegalEntityIndexID"] = response.Info.Get<Guid?>("LegalEntityID"); ;
                cardModel.Card.Sections["PnrOrderUK"].Fields["LegalEntityIndexIdx"] = response.Info.Get<string>("LegalEntityIdx");
            }
        }

        public override Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            // Создание карточки
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата регистрации текущей датой
                if (cardModel.Controls.TryGet("RegistrationDate", out IControlViewModel projectDateControl))
                {
                    ((DateTimeViewModel)projectDateControl).SelectedDate = DateTime.Now;
                }

                // начальная инициализация поля Подразделение по автору
                SetUserDepartmentInfo((Guid?)card.Sections["DocumentCommonInfo"].Fields["AuthorID"], cardModel);
            }

            context.Card.Sections["PnrOrderUK"].FieldChanged += (s, e) =>
            {
                // Организация ГК Пионер: заполнение Индекс ЮЛ + изменение Номера
                if (e.FieldName == "OrganizationID" && e.FieldValue != null)
                {
                    SetLegalEntityIndex((Guid)e.FieldValue, cardModel);
                }
            };

            return base.Initialized(context);
        }
    }
}

File: WebClient SDK/src/PnrScripts/UI/PnrUKOrderUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid, createTypedField, DotNetType } from 'tessa/platform';
import { CardStoreMode } from 'tessa/cards';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';
import { CardRequest, CardService } from 'tessa/cards/service';
import { DateTimeViewModel } from 'tessa/ui/cards/controls';
import Moment from 'moment';

export class PnrUKOrderUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrOrderUKTypeID)) {
      return;
    }
    let PnrOrderUK = context.card.sections.tryGet("PnrOrderUK");
    if (!PnrOrderUK) return;
    // Создание карточки
    if (context.card.storeMode == CardStoreMode.Insert)
    {
      // Заполнить поле Дата регистрации текущей датой
      const regDateControl = context.model.controls.get('RegistrationDate') as DateTimeViewModel;
      if (regDateControl) {
        regDateControl.selectedDate = Moment(Date.now());
      }
      let dci = context.card.sections.tryGet("DocumentCommonInfo");
      if (dci != null) {
        // начальная инициализация поля Подразделение по автору
        await this.SetUserDepartmentInfo(dci.fields.get('AuthorID'), context);
      }
    }

    PnrOrderUK.fields.fieldChanged.add(async (e) => {
      // Организация ГК Пионер: заполнение Индекс ЮЛ + изменение Номера
      if (e.fieldName == "OrganizationID" && e.fieldValue != null)
      {
        await this.SetLegalEntityIndex(e.fieldValue, context);
      }
    });
  }

  private async SetUserDepartmentInfo(authorID, context: ICardUIExtensionContext)
  {
    let DepartmentID;
    let DepartmentName;
    let DepartmentIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetUserDepartmentInfoRequestTypeID;
        request.info["authorID"] = createTypedField(authorID, DotNetType.Guid);

    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
      DepartmentID = response.info.DepartmentID;
      DepartmentName = response.info.Name;
      DepartmentIdx = response.info.Index;
    }

    let PnrOrderUK = context.card.sections.tryGet("PnrOrderUK");
    if (DepartmentID != null && DepartmentName != null && DepartmentIdx != null && PnrOrderUK != null) {
      PnrOrderUK.fields.set('DepartmentID', createTypedField(DepartmentID.$value, DotNetType.Guid));
      PnrOrderUK.fields.set('DepartmentName', createTypedField(DepartmentName.$value, DotNetType.String));
      PnrOrderUK.fields.set('DepartmentIdx', createTypedField(DepartmentIdx.$value, DotNetType.String));
    }
  }

  private async SetLegalEntityIndex(organizationID, context: ICardUIExtensionContext)
  {
    // request к организации получить индекс ЮЛ
    let LegalEntityIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetIndexLegalEntityRequestTypeID;
        request.info["organizationID"] = createTypedField(organizationID, DotNetType.Guid);
    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
        LegalEntityIdx = response.info.LegalEntityIdx;
    }
    let PnrOrderUK = context.card.sections.tryGet("PnrOrderUK");
    if (LegalEntityIdx != null && PnrOrderUK != null) {
      PnrOrderUK.fields.set('LegalEntityIndexIdx', createTypedField(LegalEntityIdx.$value, DotNetType.String))
    }
  }
}